### ISSUE TITLE
Publish Semantic Version Policy

### ISSUE THREAD

**Issue by ntwb** [stylelint/stylelint#1804]
This stems from [a conversation](https://github.com/stylelint/eslint-config-stylelint/commit/e8ac8ffe761345771f69ef21e1c3cce309b6070a) just had regarding the release of `eslint-config-stylelint`, the conversation is primarily about whether it should have been released as a minor i.e. `3.1.0` release or a major i.e. `4.0.0`.

stylelint should document and adopt an organisation wide semantic versioning policy

Rather than reinventing the wheel, I'd propose adopting ESLint's semantic versioning policy, with broad adoption and widely accepted throughout the ESLint community.

 • ESLint have semver policy: https://github.com/eslint/eslint#semantic-versioning-policy

• ESLint also has a _commit message format_ policy: I can't find the policy, but it is mentioned [here](http://eslint.org/docs/developer-guide/contributing/pull-requests#updating-the-commit-message) as a requirement, you can see examples of the polic in action by from viewing ESLint's [commits here](https://github.com/eslint/eslint/commits/master)

Do you see any issues with ESLint's policy that might will not work for stylelint?

cc @stylelint/core

**Comment by davidtheclark** [comment]
@ntwb The ESLint semver policy looks good to me, looks like it will work for us.

I'm not particularly interested in a commit message policy. My impression is that ESLint does that mostly because they auto-generate a changelog from the messages. I think the way we've been working with the changelog has been ok, without the additional burden of commit format enforcement.

**Comment by jeddy3** [comment]
> The ESLint semver policy looks good to me, looks like it will work for us.

Sgtm. I think the only change to how we currently do things is:

> Minor release (might break your lint build)
> -- A bug fix in a rule that results in ESLint reporting more errors.

We've tended to treat those as breaking. Happy to see them consistently treated a minor though.

The policy of major release for config changes e.g.

> Major release (likely to break your lint build)
> -- `eslint:recommended` is updated.

Is what we do at the moment with `stylelint-config-standard`.

> I think the way we've been working with the changelog has been ok, without the additional burden of commit format enforcement.

Also sgtm.

**Comment by ntwb** [comment]
> Minor release (might break your lint build)
> -- A bug fix in a rule that results in ESLint reporting more errors.
> > We've tended to treat those as breaking. Happy to see them consistently treated a minor though.

We may need to give the community a bit of a heads up that we are going to change/tweak our semantic versioning policy, even as much as we don't actually change our current policy until the `next-major` milestone, i.e. `8.0.0`

> I think the way we've been working with the changelog has been ok, without the additional burden of commit format enforcement.

Agreed, I suppose that's our policy and it works quite well I think 👍

**Comment by ntwb** [comment]
Researching a couple of past issues we've had or had reported regarding stylelint's semver policy:

> Via https://github.com/stylelint/stylelint/issues/1236
> _"6.3.3 contains #1205 which contains declaration-block-no-ignored-properties improvements, but it's a breaking change, because in 6.3.2 I can have no errors, but i will get it in 6.3.3. And it can break CI builds."_

The above was released in a _patch_ release `6.3.3`, based on the proposed semver policy this would be a _minor_ release per `A bug fix in a rule that results in ESLint reporting more errors.` We acknowledged that this shouldn't have been a _patch_ release in the original report.

> Via #869 
> > _"We were using 4.4.0, and when v 4.5.0 got releases today, our build broke, as we have it setup to step up automatically on minor releases."_

In this case with the proposed semver policy we potentially will break builds due to these _minor_ policies:
* `A bug fix in a rule that results in ESLint reporting more errors.`
* `An existing rule is deprecated.`

Per the ESLint semver policy:
> _"According to our policy, any minor update may report more errors than the previous release (ex: from a bug fix). As such, we recommend using the tilde (~) in package.json e.g. `"eslint": "~3.1.0"` to guarantee the results of your builds."_

To sum up, the two items mentioned above that are part of the _minor_ policy would be what I expect most issues to arise from, that said, recommending the change to using the tilde (`~`) in `package.json`, e.g. `"stylelint": "~7.1.0"` should mitigate the majority of those issues.

> Major release (likely to break your lint build)
> -- `eslint:recommended` is updated.
> >Is what we do at the moment with `stylelint-config-standard`.

With what I've outlined above I think we'd continue the same practice, for the most part `stylelint-config-standard` would continue to be released as a _major_ version

**Comment by jeddy3** [comment]
> We may need to give the community a bit of a heads up that we are going to change/tweak our semantic versioning policy

Sgtm. I'll label and milestone this up accordingly.

> Researching a couple of past issues we've had or had reported regarding stylelint's semver policy

Thanks for doing that. It all looks good to me.

>  that said, recommending the change to using the tilde (`~`) in `package.json`, e.g. `"stylelint": "~7.1.0"` should mitigate the majority of those issues.

Agreed.

> With what I've outlined above I think we'd continue the same practice, for the most part `stylelint-config-standard` would continue to be released as a _major_ version

Agreed.

**Comment by davidtheclark** [comment]
👍

**Comment by davidtheclark** [comment]
@jeddy3 Why did you mark this as v8? Why not implement it now?

**Comment by jeddy3** [comment]
@ntwb [said](https://github.com/stylelint/stylelint/issues/1804#issuecomment-240270369):

> We may need to give the community a bit of a heads up that we are going to change/tweak our semantic versioning policy, even as much as we don't actually change our current policy until the next-major milestone, i.e. `8.0.0`

I think the rationale is that tweaking our policy before `8.0.0` could lead to *unexpected* broken builds. We can use the `8.0.0` release to *notify everyone* of the change in policy, and the recommendation that update their version range to a tilde.

I guess the validity of that approach is up for discussion. Also, there's no harm in just throwing out `8.0.0` now with the policy change...

**Comment by davidtheclark** [comment]
Sounds good. I forgot about that rationale, my mistake.

**Comment by jeddy3** [comment]
Done in https://github.com/stylelint/stylelint/issues/1865
